block-library/tabs/blocks/tabs/edit.js

```diff
@@ -0,0 +1,219 @@
+		} );
+	}, [ innerBlocks, setAttributes, currentActiveTabInstanceId ] );
+
+	/**
+	 * @function updateTabLabel
+	 *
+	 * @description Dispatch action to matching tab to update its label
+	 *
+	 * @param {string} tabLabel
+	 * @param {string} tabClientId
+	 */
+	const updateTabLabel = ( tabLabel, tabClientId ) => {
+		dispatch.updateBlockAttributes( tabClientId, { tabLabel } );
+	};
+
+	/**
+	 * @function addNewTab
+	 *
+	 * @description adds a new tab and an InnerBlock to match
+	 *
+	 * @param {number} positionToAdd
+	 */
+	const addNewTab = ( positionToAdd = innerBlocks.length ) => {
+		// create block
+		const newTab = createBlock( 'tribe/tab' );
+
+		// add new tab
+		dispatch
+			.insertBlock( newTab, positionToAdd, clientId, true )
+			.then( () => {
+				// dispatch will return us a promise which we can use to set our new active tab instanceId
+				const newInstanceId =
+					select.getBlocks( clientId )[ positionToAdd ].attributes
+						.blockId;
+
+				// set new tab as active
+				setAttributes( {
+					currentActiveTabInstanceId: newInstanceId,
+				} );
+			} );
+	};
+
+	/**
+	 * @function deleteTab
+	 *
+	 * @description handles removing tab & InnerBlock based on index
+	 *
+	 * @param {number} index
+	 * @param {string} tabInstanceId
+	 * @param {string} tabClientId
+	 */
+	const deleteTab = ( index, tabInstanceId, tabClientId ) => {
+		// remove block from InnerBlocks
+		removeBlocks( tabClientId );
+
+		// Fetch new inner blocks
+		const newInnerBlocks = select.getBlocks( clientId );
+
+		// Add a new tab if we've deleted the last one
+		if ( newInnerBlocks.length === 0 ) {
+			addNewTab( newInnerBlocks.length );
+
+			return;
+		}
+
+		// decide which block should be the new selected block
+		let newActiveTabInstanceId = currentActiveTabInstanceId;
+
+		if ( newActiveTabInstanceId === tabInstanceId ) {
+			// if we want the first block, show new "first block", any other tab, use the "next" tab
+			newActiveTabInstanceId =
+				index === 0
+					? newInnerBlocks[ index ].attributes.blockId
+					: newInnerBlocks[ index - 1 ].attributes.blockId;
+		}
+
+		setAttributes( {
+			currentActiveTabInstanceId: newActiveTabInstanceId,
+		} );
+	};
+
+	// setup inner block props and add classname to wrapper
+	const innerBlocksProps = useInnerBlocksProps(
+		{
+			className: 'wp-block-tribe-tabs__tab-content',
+		},
+		{
+			allowedBlocks: [ 'tribe/tab' ],
+			template: [ [ 'tribe/tab' ] ],
+			renderAppender: false,
+		}
+	);
+
+	// set new tab state when innerBlocks or currentActiveTabInstanceId changes
+	useEffect( () => {
+		const data = innerBlocks.map( ( tab ) => {
+			return {
+				clientId: tab.clientId,
+				id: tab.attributes.blockId,
+				buttonId: 'button-' + tab.attributes.blockId,
+				label: tab.attributes.tabLabel,
+				isActive: currentActiveTabInstanceId === tab.attributes.blockId,
+			};
+		} );
+
+		setAttributes( {
+			tabs: data,
+		} );
+	}, [ innerBlocks, currentActiveTabInstanceId, setAttributes ] );
```

Comment:
Looks like this comment got removed from my first pass, note sure what happened.

I was wondering if this entire effect is necessary. Instead of creating a new state here to hold the tabs, can we just use `innerBlocks` directly? If this component re-renders every time the inner blocks change, I think that should work fine and save us some steps.

Reply:
I worked on this for a couple hours this morning. I realized that I had probably already tried the method you proposed here and got stuck. This is because while `innerBlocks` is available and would work fine in the `edit` function, it's not available in the `save` function. This means we have no way to create the tabs within the `save` function if we take that route. I believe this was the only way I could figure out how to get data into the `save` function.

I'm going to add a todo item to review optimizations for this block at a later time, just so we can get a working version of this block into the repo.

Reply:
Cool, thank you for taking a look at that. WordPress makes it challenging to do stick to sort of standard React workflows because of saving things to the database and rendering them outside of React on the FE.